Allow different param names in different methods with same path scheme

ortyomka opened the pull request:
Relates to issues  https://github.com/labstack/echo/issues/1726 and https://github.com/labstack/echo/issues/2201

Parameters and paths are now separated by methods within the same node. 
Add new method Context, which store path, parameters, and handler.
Add test from https://github.com/labstack/echo/issues/1726. 

CC @aldas

ortyomka commented:
@aldas could you run tests again? Added small fix

ortyomka commented:
I forgot about middleware tests. Fixed

@aldas

ortyomka commented:
Found a quicker and easier way for the new structure. In `methodHandler`, I use `Context` instead of `Handler` and don't store map.

old - original
new - map with contexts
new2 - `methodHandler` with contexts
```
name \ time/op                       old.txt      new.txt      new2.txt
EchoStaticRoutes-4                   18.1µs ± 0%  18.4µs ± 0%  17.4µs ± 0%
EchoStaticRoutesMisses-4             17.7µs ± 0%  20.1µs ± 0%  17.5µs ± 0%
EchoGitHubAPI-4                      31.9µs ± 0%  34.0µs ± 0%  30.5µs ± 0%
EchoGitHubAPIMisses-4                40.7µs ± 0%  33.5µs ± 0%  31.1µs ± 0%
EchoParseAPI-4                       3.11µs ± 0%  2.30µs ± 0%  2.03µs ± 0%
RouterStaticRoutes-4                 14.0µs ± 0%  15.2µs ± 0%  13.5µs ± 0%
RouterStaticRoutesMisses-4            491ns ± 0%   492ns ± 0%   466ns ± 0%
RouterGitHubAPI-4                    25.3µs ± 0%  27.4µs ± 0%  23.7µs ± 0%
RouterGitHubAPIMisses-4               576ns ± 0%   594ns ± 0%   568ns ± 0%
RouterParseAPI-4                     1.28µs ± 0%  1.47µs ± 0%  1.23µs ± 0%
RouterParseAPIMisses-4                304ns ± 0%   363ns ± 0%   312ns ± 0%
RouterGooglePlusAPI-4                 854ns ± 0%  1137ns ± 0%   823ns ± 0%
RouterGooglePlusAPIMisses-4           466ns ± 0%   683ns ± 0%   477ns ± 0%
RouterParamsAndAnyAPI-4              1.96µs ± 0%  2.54µs ± 0%  2.03µs ± 0%
```

original vs current
```
name                                 old time/op    new time/op    delta
EchoStaticRoutes-4                     17.2µs ± 2%    17.3µs ± 2%    ~     (p=0.421 n=5+5)
EchoStaticRoutesMisses-4               17.0µs ± 2%    17.3µs ± 2%    ~     (p=0.151 n=5+5)
EchoGitHubAPI-4                        30.9µs ± 0%    31.3µs ± 3%    ~     (p=0.690 n=5+5)
EchoGitHubAPIMisses-4                  31.5µs ± 4%    31.5µs ± 5%    ~     (p=1.000 n=5+5)
EchoParseAPI-4                         2.04µs ± 3%    2.05µs ± 2%    ~     (p=0.841 n=5+5)
RouterStaticRoutes-4                   13.5µs ± 3%    13.6µs ± 2%    ~     (p=0.095 n=5+5)
RouterStaticRoutesMisses-4              474ns ± 1%     469ns ± 0%  -1.01%  (p=0.008 n=5+5)
RouterGitHubAPI-4                      23.9µs ± 1%    23.9µs ± 2%    ~     (p=0.841 n=5+5)
RouterGitHubAPIMisses-4                 577ns ± 0%     580ns ± 3%    ~     (p=0.730 n=4+5)
RouterParseAPI-4                       1.24µs ± 1%    1.24µs ± 1%    ~     (p=0.222 n=5+5)
RouterParseAPIMisses-4                  306ns ± 3%     300ns ± 0%  -1.84%  (p=0.008 n=5+5)
RouterGooglePlusAPI-4                   831ns ± 1%     806ns ± 1%  -2.93%  (p=0.008 n=5+5)
RouterGooglePlusAPIMisses-4             468ns ± 0%     462ns ± 0%  -1.29%  (p=0.008 n=5+5)
RouterParamsAndAnyAPI-4                1.94µs ± 0%    1.93µs ± 1%    ~     (p=0.071 n=5+5)
```

cc @aldas, ready for review

aldas reviewed (commented):
I do not understand why this new struct `methodContext` is necessary in this PR context? I do not see it adding any functional or performance value.  Changing things from `HandlerFunc` to pointer seems unnecessary as function types are nillable.

ortyomka commented:
> I do not understand why this new struct `methodContext` is necessary in this PR context? I do not see it adding any functional or performance value. Changing things from `HandlerFunc` to pointer seems unnecessary as function types are nillable.

Agree, I will add functionality

ortyomka commented:
Added functionality, to store different paths and parameters in different methods

Done @aldas

ortyomka commented:
@aldas , if you have time, please review PR

aldas commented on router.go:
```diff
@@ -569,7 +579,11 @@ func (r *Router) Find(method, path string, c Context) {
 			// If any node is found, use remaining path for paramValues
 			currentNode = child
-			paramValues[len(currentNode.pnames)-1] = search
+			if m := currentNode.findHandler(method); m != nil {
```
I think better approach would be to compute parameter count as a field on node. Not that calling `currentNode.findHandler(method)` on each any handler is super slow but it is still unnecessary most of this work can be computed beforehand.  You can check how in `v5` branch this is done. See here for usage of [`node.paramsCount`](https://github.com/labstack/echo/blob/v5_alpha/router.go).

ortyomka replied:
Good approach, added

ortyomka commented on router.go:
```diff
@@ -13,14 +13,17 @@ type (
 		routes map[string]*Route
 		echo   *Echo
 	}
+	methodContext struct {
```
Renamed to routeMethod

aldas commented on router.go:
```diff
@@ -604,22 +606,23 @@ func (r *Router) Find(method, path string, c Context) {
 		// so we can send http.StatusMethodNotAllowed (405) instead of http.StatusNotFound (404)
 		currentNode = previousBestMatchNode
 
+		ctx.path = path
```
See comment below. This is incorrect as `path` must be path of the route not path of the request.

ortyomka replied:
Fixed

aldas commented on router.go:
```diff
@@ -604,22 +606,23 @@ func (r *Router) Find(method, path string, c Context) {
 		}
 	}
-	ctx.path = currentNode.ppath
-	ctx.pnames = currentNode.pnames
```
I do not think that moving these assignments is a good thing. Although `c.Reset` will reset context in `e.ServeHTTP` it is more foolproof to make sure that every codepath will assign something in router to these values. so doing these here is better than relying some if branch somewhere be smart enough to know all corner cases.  

Consider this testcase which will fail due after second  `r.Find`  param names still contains old values from previous request and path value is set incorrectly to request path. If you check  `context.Path()` and `context.ParamValues()` you will see that their result depends directly on `pnames` contents/size.
```go
func TestRouterDifferentParamsInPath(t *testing.T) {
	e := New()
	r := e.router
	r.Add(http.MethodPut, "/*", func(Context) error {
		return nil
	})
	r.Add(http.MethodPut, "/users/:vid/files/:gid", func(Context) error {
		return nil
	})
	r.Add(http.MethodGet, "/users/:uid/files/:fid", func(Context) error {
		return nil
	})
	c := e.NewContext(nil, nil).(*context)

	r.Find(http.MethodGet, "/users/1/files/2", c)
	assert.Equal(t, c.ParamNames(), []string{"uid", "fid"})
	assert.Equal(t, "1", c.Param("uid"))
	assert.Equal(t, "2", c.Param("fid"))

	r.Find(http.MethodGet, "/users/1/shouldBacktrackToFirstAnyRouteAnd405", c)
	assert.Equal(t, "/*", c.Path()) // we should have (partially by 405) matched route with path `/*` by backtracking logic

	r.Find(http.MethodPut, "/users/3/files/4", c)
	assert.Equal(t, c.ParamNames(), []string{"vid", "gid"})
	assert.Equal(t, "3", c.Param("vid"))
	assert.Equal(t, "4", c.Param("gid"))
}
```

ortyomka replied:
Fixed

ortyomka commented:
@aldas 
I corrected comments, but a question remains. Do we need to have parameters when we have 405?

I removed them [now](https://github.com/labstack/echo/pull/2209/files#diff-d1e1aef472c68045d1f3c4f68a932e458699037baa4e68d5ca07629d4792bee0R629), but if we need them, what names should they have?

ortyomka commented:
I would like to finish this PR by Tuesday(12.07) inclusive. 

If your schedule allows it, I'm willing to fix comments ASAP. 

CC @aldas

aldas reviewed (approved):
LGTM

I think this PR is good enough for current state. Params for 405 etc are things we/I can address later when there is a decision what the behavior should be. These questions are not necessary to be solved with this change.  

Thank you for PR.
